internal/uvm/scsi.go

```diff
@@ -323,9 +347,41 @@ func (uvm *UtilityVM) addSCSIActual(ctx context.Context, hostPath, uvmPath, atta
 	}
 
 	if err := uvm.modify(ctx, SCSIModification); err != nil {
-		return nil, fmt.Errorf("uvm::AddSCSI: failed to modify utility VM configuration: %s", err)
+		return fmt.Errorf("uvm::AddSCSI: failed to modify utility VM configuration: %s", err)
 	}
-	return sm, nil
+	return nil
+}
+
+// allocateSCSIDevice grants vm access to hostpath and increments the ref count of an existing scsi
+// device or allocates a new one if not already present.
+// Returns the resulting *SCSIMount, a bool indicating if the scsi device was already present,
+// and error if any.
+func (uvm *UtilityVM) allocateSCSIDevice(ctx context.Context, hostPath, uvmPath string, vmAccess VMAccessType) (*SCSIMount, bool, error) {
+	existed := false
+	// Ensure the utility VM has access
+	err := uvm.grantAccess(ctx, hostPath, vmAccess)
+	if err != nil {
+		return nil, false, err
+	}
+	// We must hold the lock throughout the lookup (findSCSIAttachment) until
+	// after the possible allocation (allocateSCSISlot) has been completed to ensure
+	// there isn't a race condition for it being attached by another thread between
+	// these two operations. All failure paths between these two must release
+	// the lock.
+	uvm.m.Lock()
+	defer uvm.m.Unlock()
+	if sm, err := uvm.findSCSIAttachment(ctx, hostPath); err == nil {
+		existed = true
```

Comment:
It looks kind of awkward to be tracking this as a separate value that we need to assign to before returning. Was this done to make it clearer what the value indicates? I think it's fine to just return `true` or `false` inline in the return statement, rather than using a separate variable.

Reply:
Yeah the goal was to make it clearer, but if this doesn't help I'm fine taking it out :)